File: Python3/158. Read N Characters Given read4 II - Call Multiple Times.py
# The read4 API is already defined for you.
# def read4(buf4: List[str]) -> int:

class Solution:
    def __init__(self):
        self.copied = 0
        self.buf4 = [' ']*4
    
    def read(self, buf: List[str], n: int) -> int:
        
        total = 0
        read = 4
        
        # copy the remaining characters from self.last to buf
        if self.copied:
            buf[:self.copied] = self.buf4[:self.copied]
            total += self.copied
            self.copied = 0
        
        while read and total<n:
            buf4 = [' ']*4
            read = read4(buf4)
            buf[total:total+read] = buf4[:read]
            total += read
        
        if total > n:
            self.buf4[:total-n] = buf[n:total]
            self.copied = total-n
        
        return min(total, n)

File: Java/158. Read N Characters Given read4 II - Call Multiple Times.java
/**
 * The read4 API is defined in the parent class Reader4.
 *     int read4(char[] buf4); 
 */

public class Solution extends Reader4 {
    /**
     * @param buf Destination buffer
     * @param n   Number of characters to read
     * @return    The number of actual characters read
     */
    
    int copied = 0;
    char[] legacy = new char[4];
    
    public int read(char[] buf, int n) {
        int total = 0;
        int read = 4;
            
        while(copied>0){
            //System.out.print("legacy = ");
            //System.out.println(legacy);
            buf[total] = legacy[total];
            total++;
            copied--;
            //System.out.print("buf[total] = ");
            //System.out.println(buf[total]);
        }
        
        
        while(read>0 && total<n){
            char[] buf4 = new char[4];
            read = read4(buf4);
            
            for(int i=0; i<read; i++){
                buf[total] = buf4[i];
                //System.out.print("buf[total+i] = ");
                //System.out.println(buf[total+i]);
                total++;
            }
        }
        //System.out.println("total = " + total);
        
        for(int i=0; i<(total-n); i++){
            legacy[i] = buf[n+i];
            
            //System.out.print("buf[n+i] = ");
            //System.out.println(buf[n+i]);
        }
        copied = total-n;
        
        return Math.min(total, n);
    }
}